src/main/scala/org/scalameter/japi/JBench.scala

```diff
@@ -0,0 +1,271 @@
+      setScope(clazz, scopeCtxs, scope, mapping.map(kv => (kv._1.tail, kv._2)))
+    }
+  }
+
+  /** Sets given scope.
+   *
+   *  It sets declared scopes recursively,
+   *  from top to the bottom where setup items are created.
+   */
+  private def setScope(clazz: Class[_],
+    scopeCtxs: Map[Seq[String], Context], name: String,
+    scopedSetups: Seq[(Seq[String], Seq[Method])]): Unit = {
+    val scope = Scope(name, setupzipper.value.current.context)
+    scope config {
+      scopeCtxs.getOrElse(
+        (scope.name :: scope.context(Key.dsl.scope)).reverse, Context.empty
+      )
+    } in {
+      scopedSetups.groupBy(_._1.headOption).foreach {
+        case (None, newMapping) =>
+          newMapping.flatMap(_._2).foreach(m => setSetup(clazz, m))
+        case (Some(newScope), newMapping) =>
+          setScope(
+            clazz, scopeCtxs, newScope, newMapping.map(kv => (kv._1.tail, kv._2))
+          )
+      }
+    }
+  }
+
+  /** Sets setup for a benchmark snippet.
+   */
+  private def setSetup(cl: Class[_], m: Method) {
+    val additionalContext = Option(m.getAnnotation(classOf[ctx]))
+      .map(a => getField[Context](cl, a.value()))
+      .getOrElse(Context.empty)
+
+    val gen = Option(m.getAnnotation(classOf[gen]))
+      .map(a => getField[Gen[AnyRef]](cl, a.value()))
+      .getOrElse(sys.error("Each benchmark method should be annotated with 'gen'."))
+
+    val setupBeforeAll= getNoArgMethod(m, classOf[setupBeforeAll], cl)
+      .asInstanceOf[Option[() => Unit]]
+
+    val teardownAfterAll = getNoArgMethod(m, classOf[teardownAfterAll], cl)
+      .asInstanceOf[Option[() => Unit]]
+
+    val setup: Option[AnyRef => Any] = getOneArgMethod(m, classOf[setup], cl)
+
+    val teardown: Option[AnyRef => Any] = getOneArgMethod(m, classOf[teardown], cl)
+
+    val warmup: Option[() => Any] = getNoArgMethod(m, classOf[warmup], cl)
+
+    val snippet: AnyRef => Any = {
+      m.setAccessible(true)
+      val sm = new SerializableMethod(m)
+      v => sm.invokeA(this, v)
+    }
+
+    val curveName = Option(m.getAnnotation(classOf[curve]))
+      .map(_.value()).getOrElse(m.getName)
+
+    val context = setupzipper.value.current.context +
+      (Key.dsl.curve -> curveName) ++ additionalContext
+
+    val benchmark = Setup[AnyRef](
+      context = context,
+      gen = gen,
+      setupbeforeall = setupBeforeAll,
+      teardownafterall = teardownAfterAll,
+      setup = setup,
+      teardown = teardown,
+      customwarmup = warmup,
+      snippet = snippet,
+      customExecutor = executor
+    )
+
+    setupzipper.value = setupzipper.value.addItem(benchmark)
+  }
+
+
+  /** Gets value from a field or no-arg method.
+   */
+  private def getField[T](selfClass: Class[_], name: String): T = {
+    Try(selfClass.getField(name).get(this).asInstanceOf[T]) orElse
+      Try(selfClass.getMethod(name).invoke(this).asInstanceOf[T]) getOrElse
+      sys.error(throw new NoSuchFieldException(name))
```

Comment:
Since our clients will be looking at this error, it would be helpful to expand the error message a bit - say that method X in benchmark class Y is referring to non-existent field or 0-arg method Z.

Btw, `sys.error` takes a string - this just throws an exception without invoking `sys.error`.